File: frontend/src/api/clipboardApi.js
import { apiData, jsonRequestOptions } from './http.js'

function connectionPath(clientId, suffix) {
  return `/api/connections/${encodeURIComponent(clientId)}/clipboard/${suffix}`
}

export function getClipboardCapabilities(clientId) {
  return apiData(connectionPath(clientId, 'capabilities'), {}, {})
}

export function getRemoteClipboard(clientId) {
  return apiData(connectionPath(clientId, 'get'), { method: 'POST' }, {})
}

export function setRemoteClipboardText(clientId, text) {
  return apiData(
    connectionPath(clientId, 'set-text'),
    jsonRequestOptions('POST', { text }),
    {},
  )
}

export function setRemoteClipboardImage(clientId, file) {
  const form = new FormData()
  form.append('file', file, file?.name || 'clipboard.png')
  return apiData(connectionPath(clientId, 'set-image'), { method: 'POST', body: form }, {})
}

export function setRemoteClipboardArtifactFile(clientId, artifactId) {
  return apiData(
    connectionPath(clientId, 'set-artifact-file'),
    jsonRequestOptions('POST', { artifact_id: artifactId }),
    {},
  )
}

export function setRemoteClipboardFiles(clientId, items = []) {
  const form = new FormData()
  const uploads = []
  const entries = []
  const roots = []

  items.forEach((item) => {
    const kind = String(item?.kind || 'file')
    const rootName = String(item?.name || item?.file?.name || '').trim()
    if (!rootName) return

    roots.push(rootName)

    if (kind === 'directory') {
      const directories = Array.isArray(item.directories) ? item.directories : []
      const directoryFiles = Array.isArray(item.files) ? item.files : []

      directories.forEach((path) => {
        entries.push({ type: 'directory', path })
      })
      directoryFiles.forEach((entry) => {
        const file = entry?.file
        const path = String(entry?.relativePath || '').trim()
        if (!file || !path) return
        const uploadIndex = uploads.length
        uploads.push(file)
        entries.push({ type: 'file', path, upload_index: uploadIndex })
      })
      return
    }

    if (!item?.file) return
    const uploadIndex = uploads.length
    uploads.push(item.file)
    entries.push({ type: 'file', path: rootName, upload_index: uploadIndex })
  })

  uploads.forEach((file) => form.append('files', file, file?.name || 'clipboard_file'))
  form.append('manifest', JSON.stringify({ roots, entries }))
  return apiData(connectionPath(clientId, 'set-files'), { method: 'POST', body: form }, {})
}

export function downloadRemoteClipboardFile(clientId, path, headers = {}) {
  const url = `/api/connections/${encodeURIComponent(clientId)}/remote-files/download?path=${encodeURIComponent(path)}`
  return apiData(url, { method: 'POST', headers: { ...(headers || {}) } }, {})
}

export function downloadRemoteClipboardDirectory(clientId, path, headers = {}) {
  const url = `/api/connections/${encodeURIComponent(clientId)}/remote-files/download-zip`
  return apiData(
    url,
    jsonRequestOptions('POST', { paths: [path], archive_name: '' }, headers),
    {},
  )
}

File: backend/server/application/web/clipboard_api.py
import json
import os


class WebClipboardApi:
    def __init__(self, clipboard_session_service, artifact_service):
        self.clipboard_session_service = clipboard_session_service
        self.artifact_service = artifact_service

    def get_capabilities(self, client_id: str) -> dict:
        payload = self.clipboard_session_service.get_capabilities(client_id)
        return payload.get('capabilities') if isinstance(payload.get('capabilities'), dict) else {}

    def get_clipboard(self, client_id: str) -> dict:
        return self.clipboard_session_service.get_clipboard(client_id)

    def set_text(self, client_id: str, text: str) -> dict:
        return self.clipboard_session_service.set_clipboard(
            client_id,
            {'kind': 'text', 'text': str(text or '')},
        )

    def set_image(self, client_id: str, upload) -> dict:
        temp_paths = []
        try:
            item, temp_path = self._stage_upload(upload)
            temp_paths.append(temp_path)
            return self.clipboard_session_service.set_clipboard(
                client_id,
                {'kind': 'image', 'items': [item]},
            )
        finally:
            self._cleanup_temp_paths(temp_paths)

    def set_files(self, client_id: str, uploads: list, manifest_text: str) -> dict:
        try:
            manifest = json.loads(str(manifest_text or ''))
        except Exception as e:
            raise ValueError('clipboard file manifest is required') from e

        if not isinstance(manifest, dict):
            raise ValueError('clipboard file manifest is invalid')

        entries = manifest.get('entries') if isinstance(manifest.get('entries'), list) else []
        roots = manifest.get('roots') if isinstance(manifest.get('roots'), list) else []
        if not entries or not roots:
            raise ValueError('clipboard file manifest is empty')

        temp_paths = []
        items = []
        try:
            for entry in entries:
                if not isinstance(entry, dict):
                    continue

                entry_type = str(entry.get('type') or '').strip().lower()
                relative_path = str(entry.get('path') or '').strip()
                if not relative_path:
                    continue

                if entry_type == 'directory':
                    items.append({
                        'type': 'directory',
                        'path': relative_path,
                    })
                    continue

                if entry_type != 'file':
                    continue

                try:
                    upload_index = int(entry.get('upload_index'))
                except Exception as e:
                    raise ValueError('clipboard file upload index is invalid') from e

                if upload_index < 0 or upload_index >= len(uploads):
                    raise ValueError('clipboard file upload is missing')

                upload = uploads[upload_index]
                if not upload or not getattr(upload, 'filename', ''):
                    raise ValueError('clipboard file upload is missing')

                item, temp_path = self._stage_upload(upload)
                item.update({
                    'type': 'file',
                    'path': relative_path,
                })
                items.append(item)
                temp_paths.append(temp_path)

            if not items:
                raise ValueError('clipboard file manifest is empty')

            normalized_roots = [
                str(root or '').strip()
                for root in roots
                if str(root or '').strip()
            ]
            if not normalized_roots:
                raise ValueError('clipboard file roots are empty')

            return self._set_files_payload(client_id, items, normalized_roots)
        finally:
            self._cleanup_temp_paths(temp_paths)

    def set_artifact_file(self, client_id: str, artifact_id: str) -> dict:
        normalized_artifact_id = str(artifact_id or '').strip()
        if not normalized_artifact_id:
            raise ValueError('artifact_id is required')

        capabilities = self.get_capabilities(client_id)
        if not capabilities.get('files'):
            raise RuntimeError('Target clipboard does not support files')

        artifact = self.artifact_service.get_artifact_by_id(normalized_artifact_id)
        if str(artifact.get('artifact_type') or '').strip() != 'shared_files':
            raise ValueError('Only shared files can be sent to clipboard')

        source_path = self.artifact_service.get_artifact_file_path(normalized_artifact_id)
        display_name = str(
            artifact.get('original_name')
            or artifact.get('stored_name')
            or ''
        ).strip()
        display_name = os.path.basename(display_name) or os.path.basename(source_path)

        temp_paths = []
        try:
            temp_path, _safe_name = self.artifact_service.stage_local_file(source_path, display_name)
            temp_paths.append(temp_path)
            relative_url = self.artifact_service.build_upload_temp_download_relative_url(temp_path)
            return self._set_files_payload(
                client_id,
                [{
                    'type': 'file',
                    'path': display_name,
                    'name': display_name,
                    'url': relative_url,
                }],
                [display_name],
            )
        finally:
            self._cleanup_temp_paths(temp_paths)

    def _set_files_payload(self, client_id: str, items: list, roots: list) -> dict:
        return self.clipboard_session_service.set_clipboard(
            client_id,
            {
                'kind': 'files',
                'items': items,
                'roots': roots,
            },
        )

    def _stage_upload(self, upload) -> tuple[dict, str]:
        temp_path, safe_name = self.artifact_service.create_upload_temp_file(upload)
        relative_url = self.artifact_service.build_upload_temp_download_relative_url(temp_path)
        return {'name': safe_name, 'url': relative_url}, temp_path

    def _cleanup_temp_paths(self, paths: list[str]):
        for path in paths:
            try:
                self.artifact_service.cleanup_upload_temp_file(path)
            except Exception:
                pass
